Error 1105: should ensure all columns have the same length

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

This issue is discovered by random test:

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_ghpr_integration-copr-test/detail/tikv_ghpr_integration-copr-test/2377/pipeline/

This SQL sometimes fail due to a memory problem:

```sql
SELECT LEAST( '2012-02-21', 0, '2000-11-19 07:53:12.013663' ) AS field1, COALESCE( `col_datetime`, `col_enum`, ( ( GREATEST( '2016-05-25 18:03:50.043233', ( COALESCE( `col_blob` ) ) ) ) >= '2021-09-07' ), NULL, '14:58:30.040393' ) AS field2 FROM `table20_int_autoinc` WHERE INTERVAL( `col_float_unsigned`, `col_int`, '03:50:55.005501', ( `col_decimal_unsigned` > `col_int_key` ), '1992-04-12 12:04:33.033564' ) ORDER BY field1, field2 LIMIT 7 /* QNO 959 CON_ID 196 */ ;
```

In the test above, it fails when running with mocktikv.

You can see test data in https://github.com/tikv/copr-test/tree/master/push-down-test/prepare

### 2. What did you expect to see?

```
[2020-04-07T12:57:13.995Z] field1	field2
[2020-04-07T12:57:13.995Z] 0	0000-00-00 00:00:00
[2020-04-07T12:57:13.995Z] 0	0000-00-00 00:00:00
[2020-04-07T12:57:13.995Z] 0	0000-00-00 00:00:00
[2020-04-07T12:57:13.995Z] 0	0000-00-00 00:00:00
[2020-04-07T12:57:13.995Z] 0	0000-00-00 00:00:00
[2020-04-07T12:57:13.995Z] 0	0000-00-00 00:00:00
[2020-04-07T12:57:13.995Z] 0	0000-00-00 00:00:00
```

### 3. What did you see instead?

Error 1105: should ensure all columns have the same length

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

master

Another occurrence from https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_ghpr_integration-copr-test/detail/tikv_ghpr_integration-copr-test/3553/pipeline/ of https://github.com/tikv/tikv/pull/8466
```
[2020-08-25T11:01:21.325Z] Statement: #1567 -  SELECT STRCMP( ( `col_double_key` <=> `col_tinyint_unsigned_key` ), ( COALESCE( `col_varchar_1_key`, 613896924205940736, ( `col_int_key` > ( ( LEAST( `col_blob`, `col_enum` ) ) <=> ( `col_double_unsigned` NOT IN ( -2619124658292654080, ( '1997-05-24' LIKE ( ( ( 'asrpqaqrasisjngjaugvywtekbvahezvxxcihhakdstkwxudygixivxzpetimpnsumistumumeopraadbqujmpiqbhfjafgerliznrpxgpdlhkodjpmewmgfzgpxfikdodqftyppixekgmpyuospgksueegnafoeajdcsopomtithdxzndswkzgtbulphegfzttopmpwnlt' BETWEEN ( 31571 != '17:51:34.012966' ) AND `col_bigint_unsigned_key` ) != '1979-07-23' ) != 0 ) ), `col_decimal_unsigned`, '2033-02-16 17:01:01.015415', ( GREATEST( `col_binary_8_key`, '2030-01-13 10:57:36.017758', `col_tinyint` ) ) ) ) ) ), '1993-07-10 21:03:03.021372', '2034-06-16' ) ) ) AS field1 FROM `table10_int_autoinc` WHERE `col_bigint_key` <> ( `col_varbinary_32_key` <> '1982-01-20 00:37:32.057740' ) ORDER BY field1 LIMIT 1 /* QNO 1569 CON_ID 196 */ ;

[2020-08-25T11:01:21.325Z] NoPushDown Output: 

[2020-08-25T11:01:21.325Z] field1

[2020-08-25T11:01:21.325Z] -1

[2020-08-25T11:01:21.325Z] 

[2020-08-25T11:01:21.325Z] 

[2020-08-25T11:01:21.325Z] WithPushDown Output: 

[2020-08-25T11:01:21.325Z] Error 1105: should ensure all columns have the same length

[2020-08-25T11:01:21.325Z] 

[2020-08-25T11:01:21.325Z] 

[2020-08-25T11:01:21.325Z] NoPushDown Plan: 

[2020-08-25T11:01:21.325Z] id	estRows	task	access object	operator info

[2020-08-25T11:01:21.325Z] Projection_7	1.00	root		strcmp(cast(nulleq(push_down_test_db.table10_int_autoinc.col_double_key, cast(push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, double UNSIGNED BINARY)), var_string(20)), coalesce(push_down_test_db.table10_int_autoinc.col_varchar_1_key, 613896924205940736, cast(gt(push_down_test_db.table10_int_autoinc.col_int_key, nulleq(cast(least(push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_enum), double BINARY), cast(not(in(push_down_test_db.table10_int_autoinc.col_double_unsigned, -2.619124658292654e+18, cast(like(1997-05-24, cast(ne(ne(and(0, le(0, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key)), 1979), 0), var_string(20)), 92), double BINARY), cast(push_down_test_db.table10_int_autoinc.col_decimal_unsigned, double UNSIGNED BINARY), 2033, greatest(cast(push_down_test_db.table10_int_autoinc.col_binary_8_key, double BINARY), 2030, cast(push_down_test_db.table10_int_autoinc.col_tinyint, double BINARY)))), double BINARY))), var_string(20)), 1993-07-10 21:03:03.021372, 2034-06-16))->Column#62

[2020-08-25T11:01:21.325Z] └─Projection_17	1.00	root		push_down_test_db.table10_int_autoinc.col_binary_8_key, push_down_test_db.table10_int_autoinc.col_bigint_key, push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varchar_1_key, push_down_test_db.table10_int_autoinc.col_double_unsigned, push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_decimal_unsigned, push_down_test_db.table10_int_autoinc.col_int_key, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varbinary_32_key, push_down_test_db.table10_int_autoinc.col_double_key, push_down_test_db.table10_int_autoinc.col_enum, push_down_test_db.table10_int_autoinc.col_tinyint

[2020-08-25T11:01:21.325Z]   └─TopN_8	1.00	root		Column#63, offset:0, count:1

[2020-08-25T11:01:21.325Z]     └─Projection_18	8000.00	root		push_down_test_db.table10_int_autoinc.col_binary_8_key, push_down_test_db.table10_int_autoinc.col_bigint_key, push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varchar_1_key, push_down_test_db.table10_int_autoinc.col_double_unsigned, push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_decimal_unsigned, push_down_test_db.table10_int_autoinc.col_int_key, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varbinary_32_key, push_down_test_db.table10_int_autoinc.col_double_key, push_down_test_db.table10_int_autoinc.col_enum, push_down_test_db.table10_int_autoinc.col_tinyint, strcmp(cast(nulleq(push_down_test_db.table10_int_autoinc.col_double_key, cast(push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, double UNSIGNED BINARY)), var_string(20)), coalesce(push_down_test_db.table10_int_autoinc.col_varchar_1_key, 613896924205940736, cast(gt(push_down_test_db.table10_int_autoinc.col_int_key, nulleq(cast(least(push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_enum), double BINARY), cast(not(in(push_down_test_db.table10_int_autoinc.col_double_unsigned, -2.619124658292654e+18, cast(like(1997-05-24, cast(ne(ne(and(0, le(0, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key)), 1979), 0), var_string(20)), 92), double BINARY), cast(push_down_test_db.table10_int_autoinc.col_decimal_unsigned, double UNSIGNED BINARY), 2033, greatest(cast(push_down_test_db.table10_int_autoinc.col_binary_8_key, double BINARY), 2030, cast(push_down_test_db.table10_int_autoinc.col_tinyint, double BINARY)))), double BINARY))), var_string(20)), 1993-07-10 21:03:03.021372, 2034-06-16))->Column#63

[2020-08-25T11:01:21.325Z]       └─TableReader_13	8000.00	root		data:Selection_12

[2020-08-25T11:01:21.325Z]         └─Selection_12	8000.00	cop[tikv]		ne(push_down_test_db.table10_int_autoinc.col_bigint_key, ne(push_down_test_db.table10_int_autoinc.col_varbinary_32_key, "1982-01-20 00:37:32.057740"))

[2020-08-25T11:01:21.325Z]           └─TableFullScan_11	10000.00	cop[tikv]	table:table10_int_autoinc	keep order:false, stats:pseudo

[2020-08-25T11:01:21.325Z] 

[2020-08-25T11:01:21.325Z] 

[2020-08-25T11:01:21.325Z] WithPushDown Plan: 

[2020-08-25T11:01:21.325Z] id	estRows	task	access object	operator info

[2020-08-25T11:01:21.325Z] Projection_7	1.00	root		strcmp(cast(nulleq(push_down_test_db.table10_int_autoinc.col_double_key, cast(push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, double UNSIGNED BINARY)), var_string(20)), coalesce(push_down_test_db.table10_int_autoinc.col_varchar_1_key, 613896924205940736, cast(gt(push_down_test_db.table10_int_autoinc.col_int_key, nulleq(cast(least(push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_enum), double BINARY), cast(not(in(push_down_test_db.table10_int_autoinc.col_double_unsigned, -2.619124658292654e+18, cast(like(1997-05-24, cast(ne(ne(and(0, le(0, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key)), 1979), 0), var_string(20)), 92), double BINARY), cast(push_down_test_db.table10_int_autoinc.col_decimal_unsigned, double UNSIGNED BINARY), 2033, greatest(cast(push_down_test_db.table10_int_autoinc.col_binary_8_key, double BINARY), 2030, cast(push_down_test_db.table10_int_autoinc.col_tinyint, double BINARY)))), double BINARY))), var_string(20)), 1993-07-10 21:03:03.021372, 2034-06-16))->Column#62

[2020-08-25T11:01:21.325Z] └─Projection_17	1.00	root		push_down_test_db.table10_int_autoinc.col_binary_8_key, push_down_test_db.table10_int_autoinc.col_bigint_key, push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varchar_1_key, push_down_test_db.table10_int_autoinc.col_double_unsigned, push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_decimal_unsigned, push_down_test_db.table10_int_autoinc.col_int_key, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varbinary_32_key, push_down_test_db.table10_int_autoinc.col_double_key, push_down_test_db.table10_int_autoinc.col_enum, push_down_test_db.table10_int_autoinc.col_tinyint

[2020-08-25T11:01:21.325Z]   └─TopN_8	1.00	root		Column#63, offset:0, count:1

[2020-08-25T11:01:21.325Z]     └─Projection_18	8000.00	root		push_down_test_db.table10_int_autoinc.col_binary_8_key, push_down_test_db.table10_int_autoinc.col_bigint_key, push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varchar_1_key, push_down_test_db.table10_int_autoinc.col_double_unsigned, push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_decimal_unsigned, push_down_test_db.table10_int_autoinc.col_int_key, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key, push_down_test_db.table10_int_autoinc.col_varbinary_32_key, push_down_test_db.table10_int_autoinc.col_double_key, push_down_test_db.table10_int_autoinc.col_enum, push_down_test_db.table10_int_autoinc.col_tinyint, strcmp(cast(nulleq(push_down_test_db.table10_int_autoinc.col_double_key, cast(push_down_test_db.table10_int_autoinc.col_tinyint_unsigned_key, double UNSIGNED BINARY)), var_string(20)), coalesce(push_down_test_db.table10_int_autoinc.col_varchar_1_key, 613896924205940736, cast(gt(push_down_test_db.table10_int_autoinc.col_int_key, nulleq(cast(least(push_down_test_db.table10_int_autoinc.col_blob, push_down_test_db.table10_int_autoinc.col_enum), double BINARY), cast(not(in(push_down_test_db.table10_int_autoinc.col_double_unsigned, -2.619124658292654e+18, cast(like(1997-05-24, cast(ne(ne(and(0, le(0, push_down_test_db.table10_int_autoinc.col_bigint_unsigned_key)), 1979), 0), var_string(20)), 92), double BINARY), cast(push_down_test_db.table10_int_autoinc.col_decimal_unsigned, double UNSIGNED BINARY), 2033, greatest(cast(push_down_test_db.table10_int_autoinc.col_binary_8_key, double BINARY), 2030, cast(push_down_test_db.table10_int_autoinc.col_tinyint, double BINARY)))), double BINARY))), var_string(20)), 1993-07-10 21:03:03.021372, 2034-06-16))->Column#63

[2020-08-25T11:01:21.326Z]       └─TableReader_13	8000.00	root		data:Selection_12

[2020-08-25T11:01:21.326Z]         └─Selection_12	8000.00	cop[tikv]		ne(push_down_test_db.table10_int_autoinc.col_bigint_key, ne(push_down_test_db.table10_int_autoinc.col_varbinary_32_key, "1982-01-20 00:37:32.057740"))

[2020-08-25T11:01:21.326Z]           └─TableFullScan_11	10000.00	cop[tikv]	table:table10_int_autoinc	keep order:false, stats:pseudo

[2020-08-25T11:01:21.326Z] 

[2020-08-25T11:01:21.326Z] 

[2020-08-25T11:01:21.326Z] 

[2020-08-25T11:01:21.326Z] 2020/08/25 19:01:21 Test summary: non-matching queries: 1, success queries: 2045, skipped queries: 952

[2020-08-25T11:01:21.326Z] 2020/08/25 19:01:21 Test summary(sql/randgen-topn/3_compare_2.sql): Test case FAIL

[2020-08-25T11:01:21.326Z] 

[2020-08-25T11:01:21.582Z] + Test finished

[2020-08-25T11:01:21.582Z]   - /home/jenkins/agent/workspace/tikv_ghpr_integration-copr-test/copr-test/push-down-test/build/push_down_test_bin exit code is 2

[2020-08-25T11:01:21.582Z] make: *** [push-down-test] Error 2
```

--- a/expression/builtin_compare_vec.go
+++ b/expression/builtin_compare_vec.go
@@ -797,6 +797,7 @@ func (b *builtinGreatestStringSig) vecEvalString(input *chunk.Chunk, result *chu
 		if err := b.args[j].VecEvalString(b.ctx, input, arg); err != nil {
 			return err
 		}
+		dst.ReserveString(n)
 		for i := 0; i < n; i++ {
 			if src.IsNull(i) || arg.IsNull(i) {
 				dst.AppendNull()

--- a/expression/builtin_compare_vec.go
+++ b/expression/builtin_compare_vec.go
@@ -254,6 +254,7 @@ func (b *builtinLeastStringSig) vecEvalString(input *chunk.Chunk, result *chunk.
 	src := result
 	arg := buf1
 	dst := buf2
+	dst.ReserveString(n)
 	for j := 1; j < len(b.args); j++ {
 		if err := b.args[j].VecEvalString(b.ctx, input, arg); err != nil {
 			return err
@@ -793,11 +794,11 @@ func (b *builtinGreatestStringSig) vecEvalString(input *chunk.Chunk, result *chu
 	src := result
 	arg := buf1
 	dst := buf2
+	dst.ReserveString(n)
 	for j := 1; j < len(b.args); j++ {
 		if err := b.args[j].VecEvalString(b.ctx, input, arg); err != nil {
 			return err
 		}
-		dst.ReserveString(n)
 		for i := 0; i < n; i++ {
 			if src.IsNull(i) || arg.IsNull(i) {
 				dst.AppendNull()

--- a/util/chunk/column.go
+++ b/util/chunk/column.go
@@ -703,7 +703,7 @@ func (c *Column) MergeNulls(cols ...*Column) {
 	}
 	for _, col := range cols {
 		if c.length != col.length {
-			panic("should ensure all columns have the same length")
+			panic(fmt.Sprintf("should ensure all columns have the same length, expect %v, but got %v", c.length, col.length))
 		}
 	}
 	for _, col := range cols {

